### ISSUE TITLE
Deno fails when a package in a monorepo uses a `./*` export

### ISSUE THREAD

**Issue by michael-aldridge-pozlo** [denoland/deno#26693]
Deno version

```
$ deno --version
deno 2.0.4 (stable, release, x86_64-unknown-linux-gnu)
v8 12.9.202.13-rusty
typescript 5.6.2
```

To demonstrate the issue I created the following mono-repo

```
$ tree
.
├── a
│   ├── deno.json
│   └── src
│       └── index.ts
├── deno.json
└── index.ts
```

The root `deno.json` just specifies the `a` package

```
$ cat deno.json 
{
    "workspace": ["./a"]
}
```

The `a` package has a single export

```
$ cat a/src/index.ts 
export const foo = 42;
```

and a minimal `deno.json`

```
$ cat a/deno.json 
{
    "name": "@github-issue-example/a",
    "exports": {
        ".": "./src/index.ts",
        "./*": "./src/*"
    }
}
```

finally, the root module `index.ts` files contains a single export

```
$ cat index.ts 
import { foo } from "@github-issue-example/a/index.ts";
```

This sort of set up worked fine with node but deno is not happy about this import

```
$ deno run index.ts 
error: Relative import path "@github-issue-example/a/index.ts" not prefixed with / or ./ or ../ and not in import map from "file:///home/michael/github-issue-demos/deno-monorepo/index.ts"
    at file:///home/michael/github-issue-demos/deno-monorepo/index.ts:1:21
```

If I change the `export` map in `a/deno.json` to explicitly list `index` everything works fine, i.e.

```
$ cat a/deno.json 
{
    "name": "@github-issue-example/a",
    "exports": {
        ".": "./src/index.ts",
        "./index.ts": "./src/index.ts"
    }
}
```

This would seem to be a bug to me but it might be a feature request, I can't quite tell from the documentation. Any advise or help greatly appreciated. Thanks.

---

This issues looks similar to https://github.com/denoland/deno/issues/20513 but I thought it different enough to warrant a separate issue.

**Comment by bartlomieju** [comment]
I believe this is working as intended. `deno.json` exports expects all entry points to be listed out verbatim and doesn't support "globs" like `./*`.

@dsherret can you please confirm it?

**Comment by michael-aldridge-pozlo** [comment]
Ah okay. Thanks for the clarification.

**Comment by dsherret** [comment]
Yeah, it needs to list out the exports. There's some cases where we go from the deno.json or manifest to all the exports (ex. linting no-slow-types and `deno install`). It's a little tricky how that could work if we allowed wildcards (ex. we'd probably have to not consider files that are excluded from publish when doing this locally and remotely we'd have to look at the manifest items to figure out the modules).

I opened https://github.com/denoland/deno_config/issues/134 for now.

**Comment by SebastienGllmt** [comment]
> I believe this is working as intended. `deno.json` exports expects all entry points to be listed out verbatim and doesn't support "globs" like `./*`.

I think this should be re-considered. I understand that it makes sense not to allow this for typescript files (it's much easier to handle just a single `mod.ts` that then imports anything else needed), but that's not an option when you're using conditional exports to package assets

For example you might wants to package `./svg/*` for a UI component package, or a `./schema/*` to fetch JSON Schemas for a backend that exposes routes, etc.

I would say the glob pattern is useful for the same reason the [recent feature](https://github.com/denoland/deno/pull/26934) to package data files with `deno compile` is useful, but instead of having to use `readTextFileSync` you can now use the `import` syntax

**Comment by ry** [comment]
@SebastienGllmt noted but for now closing as not planned.

**Comment by bartlomieju** [comment]
Status update after re-checking on Deno 2.8.1.

**Repro still holds.** A `"./*": "./src/*"` wildcard export can't be imported as a subpath (e.g. `@pkg/a/index.ts`):

```
error: Import "@github-issue-example/a/index.ts" not a dependency and not in import map
```

The explicit-listing workaround (`"./index.ts": "./src/index.ts"`) works.

**This is not workspace-specific.** I reproduced it with a standalone package that has no `workspace` field at all, the rejection fires purely from parsing the `deno.json` (even on a plain `deno check`, with no subpath import):

```
Warning Invalid "exports" field in configuration file. The './*' export must only contain alphanumeric characters, underscores (_), dashes (-), dots (.), and slashes (/).
    at .../deno.json
```

So the root cause is package-level: `deno.json` `exports` doesn't support `./*` wildcards. The workspace in the original report was just the vehicle to make the package resolvable by name; the same export would fail identically for a JSR-published package. This is squarely a `publish`/`exports` matter, not workspaces.

**On @dsherret's reopen ask:**
1. "provide a diagnostic here at least" — partially addressed: a warning now exists (it didn't on 2.0.4). But the wording frames `*` as a forbidden *character* ("must only contain alphanumeric...") rather than explaining that wildcard exports must be listed explicitly / are unfurled on publish, and the actual resolution error doesn't mention the wildcard at all. Worth improving the message.
2. "a way to inline wildcard exports to specific entries on publish" — still outstanding, tracked in denoland/deno_config#134.

Removing `needs investigation` / `triage required` since the area is now confirmed; keeping `publish`.

**Comment by nathanblair** [comment]
Requiring explicit named exports for publishing deno projects makes deno functionally incompatible with proto-generated binding systems. Its not feasible to hand-maintain every exportable name and type those generate, there could be thousands, and defending the stance by suggesting creating some kind of scripted population of a barrel file from the proto types is...well...I thought deno was trying to save the web ecosystem.
